lib/secure_headers/headers/content_security_policy.rb

```diff
@@ -108,15 +109,32 @@ def build_media_type_list_directive(directive)
     def build_source_list_directive(directive)
       source_list = @config.directive_value(directive)
       if source_list != OPT_OUT && source_list && source_list.any?
-        minified_source_list = minify_source_list(directive, source_list).join(" ")
+        cleaned_source_list = clean_malformatted_sources(directive, source_list)
+        minified_source_list = minify_source_list(directive, cleaned_source_list).join(" ")
+        [symbol_to_hyphen_case(directive), minified_source_list].join(" ").strip
+      end
+    end
 
-        if minified_source_list =~ /(\n|;)/
-          Kernel.warn("#{directive} contains a #{$1} in #{minified_source_list.inspect} which will raise an error in future versions. It has been replaced with a blank space.")
+    # Private: Calculates a modified version version of source_list:
+    # - where any expression containing a deprecated character (semicolon or newline) has been split by that
+    # character, and
+    # - any empty string expression is removed.
+    def clean_malformatted_sources(directive, source_list)
+      cleaned_source_list = []
+      semicolon_warned_yet = false
+      source_list.map do |expression|
+        if expression =~ /(\n|;)/
+          if !semicolon_warned_yet
+            Kernel.warn("#{directive} contains a #{$1} in #{source_list.join(" ").inspect} which will raise an error in future versions. It has been replaced with a blank space.")
+            semicolon_warned_yet = true
+          end
+          split_expression = expression.split(/\n|;/)
+          cleaned_source_list.concat(split_expression)
+        else
+          cleaned_source_list.append(expression)
         end
-
-        escaped_source_list = minified_source_list.gsub(/[\n;]/, " ")
-        [symbol_to_hyphen_case(directive), escaped_source_list].join(" ").strip
       end
+      cleaned_source_list.select { |value| value != "" }
```

Comment:
I can’t remember if this library only works in Rails, but if it does you can use [`compact_blank`](https://edgeapi.rubyonrails.org/classes/Enumerable.html#method-i-compact_blank) here.

```suggestion
      cleaned_source_list.compact_blank
```

Otherwise, I would do something like:

```suggestion
      cleaned_source_list.reject(&:blank?)
```

Additionally, with the suggestion above returning the array, you could call this on the return value and remove the need for `cleaned_source_list`.